network-config-analyzer/NetworkLayer.py

```diff
@@ -0,0 +1,221 @@
+from bisect import insort
+from enum import Enum
+
+from ConnectionSet import ConnectionSet
+from IstioNetworkPolicy import IstioNetworkPolicy
+from NetworkPolicy import PolicyConnections
+from Peer import IpBlock, HostEP
+
+
+# TODO: add a layer for connectivity based on service type (culsterIP / LB / NodePort)? / containers ports?
+
+
+class NetworkLayerName(Enum):
+    K8s_Calico = 0
+    Istio = 1
+    Ingress = 2
+
+    def create_network_layer(self, policies):
+        if self == NetworkLayerName.K8s_Calico:
+            return K8sCalicoNetworkLayer(policies)
+        if self == NetworkLayerName.Istio:
+            return IstioNetworkLayer(policies)
+        if self == NetworkLayerName.Ingress:
+            return IngressNetworkLayer(policies)
+        return None
+
+
+class NetworkLayersContainer(dict):
+    """
+    NetworkLayersContainer contains all relevant network layers with their associated policies.
+    It is a map from NetworkLayerName to NetworkLayer object.
+    The following assumptions should always hold:
+    (1) If none of the layers have policies, the map should have the default layer (calico_k8s), without policies, only.
+    (2) If a layer has no policies, it should not be in the map (except for the case of the default layer if required)
+    """
+    default_layer = NetworkLayerName.K8s_Calico
+
+    def __init__(self):
+        initial_layers_dict = {NetworkLayersContainer.default_layer: K8sCalicoNetworkLayer([])}
+        super().__init__(initial_layers_dict)
+
+    def __getattr__(self, name):
+        return super().__getitem__(name)
+
+    def add_policy(self, policy, layer_name):
+        """
+        Adds a policy to the given layer.
+        :param NetworkPolicy.NetworkPolicy policy: the policy to add
+        :param NetworkLayerName layer_name: the relevant layer to add the policy to
+        :return: None
+        """
+        if not isinstance(layer_name, NetworkLayerName):
+            return
+        if layer_name not in self:
+            self[layer_name] = layer_name.create_network_layer([policy])
+        else:
+            self[layer_name].add_policy(policy)
+        if self.default_layer in self and not self[self.default_layer].policies_list:
+            del self[self.default_layer]
+
+    def does_contain_single_layer(self, layer_name):
+        """
+        Checks if the given layer is the only layer in the map.
+        :param NetworkLayerName layer_name: the layer to check
+        :return: True if the layer is the only layer in the map, False otherwise
+        """
+        return len(self) == 1 and list(self.keys())[0] == layer_name
+
+    def add_empty_layer(self, layer_name):
+        if layer_name not in self:
+            self[layer_name] = layer_name.create_network_layer([])
+
+
```

Comment:
This function breaks Assumption (2). Can we instead implement a static method `empty_layer_allowed_connections()`?

Reply:
done